client/consul.go

```diff
@@ -0,0 +1,186 @@
+}
+
+type ConsulClient struct {
+	client     *consul.Client
+	logger     *log.Logger
+	shutdownCh chan struct{}
+
+	trackedServices map[string]*trackedService
+}
+
+func NewConsulClient(logger *log.Logger, consulAddr string) (*ConsulClient, error) {
+	var err error
+	var c *consul.Client
+	cfg := consul.DefaultConfig()
+	cfg.Address = consulAddr
+	if c, err = consul.NewClient(cfg); err != nil {
+		return nil, err
+	}
+
+	consulClient := ConsulClient{
+		client:          c,
+		logger:          logger,
+		trackedServices: make(map[string]*trackedService),
+		shutdownCh:      make(chan struct{}),
+	}
+
+	return &consulClient, nil
+}
+
+func (c *ConsulClient) Register(task *structs.Task, allocID string) error {
+	var mErr multierror.Error
+	for _, service := range task.Services {
+		if err := c.registerService(service, task, allocID); err != nil {
+			mErr.Errors = append(mErr.Errors, err)
+		}
+		ts := &trackedService{
+			allocId: allocID,
+			task:    task,
+			service: service,
+		}
+		c.trackedServices[service.Id] = ts
+
+	}
+
+	return mErr.ErrorOrNil()
+}
+
+func (c *ConsulClient) Deregister(task *structs.Task) error {
+	var mErr multierror.Error
+	for _, service := range task.Services {
+		c.logger.Printf("[INFO] De-Registering service %v with Consul", service.Name)
+		if err := c.deregisterService(service.Id); err != nil {
+			c.logger.Printf("[ERROR] Error in de-registering service %v from Consul", service.Name)
+			mErr.Errors = append(mErr.Errors, err)
+		}
+		delete(c.trackedServices, service.Id)
+	}
+	return mErr.ErrorOrNil()
+}
+
+func (c *ConsulClient) ShutDown() {
+	close(c.shutdownCh)
+}
+
+func (c *ConsulClient) findPortAndHostForLabel(portLabel string, task *structs.Task) (string, int) {
+	for _, network := range task.Resources.Networks {
+		if p, ok := network.MapLabelToValues()[portLabel]; ok {
+			return network.IP, p
+		}
+	}
+	return "", 0
+}
+
+func (c *ConsulClient) SyncWithConsul() {
+	sync := time.After(syncInterval)
+	agent := c.client.Agent()
+
+	for {
+		select {
+		case <-sync:
+			sync = time.After(syncInterval)
+			var consulServices map[string]*consul.AgentService
+			var err error
+
+			// Get the list of the services that Consul knows about
+			if consulServices, err = agent.Services(); err != nil {
+				c.logger.Printf("[DEBUG] Error while syncing services with Consul: %v", err)
+				continue
+			}
+
+			// See if we have services that Consul doesn't know about yet.
+			// Register with Consul the services which are not registered
+			for serviceId := range c.trackedServices {
+				if _, ok := consulServices[serviceId]; !ok {
+					ts := c.trackedServices[serviceId]
+					c.registerService(ts.service, ts.task, ts.allocId)
+				}
+			}
+
+			// See if consul thinks we have some services which are not running
+			// anymore on the node. We de-register those services
+			for serviceId := range consulServices {
+				if serviceId == "consul" {
+					continue
+				}
+				if _, ok := c.trackedServices[serviceId]; !ok {
```

Comment:
It looks like this clobbers other services that may be pre-installed on the host (e.g. monitoring or logging tools installed via puppet or baked into an AMI).

Does `trackedServices` get serialized anywhere?